### ISSUE TITLE
Split orders - Incorrect calculation of carrier availability for weight ranges

### ISSUE THREAD

**Issue by Padamdam** [PrestaShop/PrestaShop#35573]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

It appears there's an error in the logic for calculating available carriers.

In the scenario of an order that needs to be split between two carriers due to different products requiring different carriers, it's essential to verify that the weight of the products being delivered falls within the defined weight ranges for each carrier.

However, the issue arises from the fact that for a particular carrier, instead of calculating the weight of only the products assigned to that carrier, it's taking ianto account the weight of the entire order.

During testing, even though this error in checking available carriers exists, it doesn't seem to affect the calculation of the delivery amount.

## Prestashop Version
1.7.8.9

## Affected Code:
In classes/Carrier.php, method getCarriersForOrder()

```
if ($shipping_method == Carrier::SHIPPING_METHOD_WEIGHT
    && (!Carrier::checkDeliveryPriceByWeight($row['id_carrier'], $cart->getTotalWeight(), $id_zone))) {
    $error[$carrier->id] = Carrier::SHIPPING_WEIGHT_EXCEPTION;
    unset($result[$k]);

    continue;
}
```

As you can see in the code block, checkDeliveryPriceByWeight tests for the total weight of the order instead of the weight of all products shipped with the carrier.

### Expected behavior

The getCarriersForOrder() method should check for carrier weight compatibility only for the products shipped by this carrier.

### Steps to reproduce

- Create two carriers (with setting Disable carrier if weight out of range)
  - CarrierA : with ranges 0 to 3kg - 10€
  - CarrierB : with ranges 0 to 3kg - 5€

- Create two products :
  - ProductA : weight 2kg - CarrierA
  - ProductB : weight 2kg - CarrierB

- Create an cart from the front office containing :
  - 1x ProductA
  - 1x ProductB

In this scenario, $cart->getTotalWeight() = 4kg, however none of the carriers allow shipping weights above 3kg.

In the checkout page, when proceeding to the carrier selection step : 

### With debug enabled
500 error appears with the following stack :
(1/1) ContextErrorExceptionNotice: Trying to access array offset on value of type null
in DeliveryOptionsFinder.php line 87

### With debug disabled
The customer is presented with an unammed carrier with the shipping amount equal to only one of the shipping carriers correct price (either 5€ or 10€ depending on the position of products in cart).

## Workaround

If we extend both the Carrier's weight ranges to allow 4kg (eg. 0kg to 5kg), then the FO shipping price is correct (15€)

### PrestaShop version(s) where the bug happened

1.7.8.7

### PHP version(s) where the bug happened

7.4.33

**Comment by paulnoelcholot** [comment]
Hello @Padamdam,

I tested on 8.1.4 and I found a even stranger issue...

https://www.loom.com/share/82724a09eea846608fc77506d72a2804?sid=d6993f1b-726c-4536-bfed-7070dab87121

We will investigate the problem and surely create an issue that encompasses the problem more broadly.

Thanks!

**Comment by Hlavtox** [comment]
@paulnoelcholot It's normal, it just doesn't work... there are some issues for it.

- It shows only one carrier instead of the two - https://github.com/PrestaShop/PrestaShop/issues/9998
- Order confirmation only shows one order instead of all of them.
- Etc.

**Comment by paulnoelcholot** [comment]
Thank you @Hlavtox for your great memory <3 

Hello @Padamdam,

We are aware of this issue, it is already in our debug backlog. This issue is a duplicate of #9998 .
To be informed when it's fixed, please subscribe to the issue mentioned above.

Thank you

**Comment by paulnoelcholot** [comment]
Duplicate of #9998

**Comment by Padamdam** [comment]
Hi @paulnoelcholot,

Thanks for the feedback and testing out on 8.1.4 !

I dont think #9998 is a duplicate of this issue as it describes another issue.

From what I understand, #9998 focuses on which carrier is presented in FO depending on the order or products in cart and the fact that only one carrier is presented.

Reading back my initial message, I didn't describe the issue correctly :

The current issue describes a case where no carrier is presented depending on the weight configurations.
The front office doesn't show any carriers in the scenario described, preventing the customer from creating an order, **contrary to what i said previously** :

> During testing, even though this error in checking available carriers exists, it doesn't seem to affect the calculation of the delivery amount.

While both issues seem related, they both require different configurations and produce different outcomes.

**Comment by Padamdam** [comment]
> I tested on 8.1.4 and I found a even stranger issue...
> 
> https://www.loom.com/share/82724a09eea846608fc77506d72a2804?sid=d6993f1b-726c-4536-bfed-7070dab87121
> 
> We will investigate the problem and surely create an issue that encompasses the problem more broadly.
> 
> Thanks!

Regarding this test, the carriers configuration are not setup to recreate the issue.

The **Out-of-range behavior** was set on **Apply the cost of the highest defined range** instead of **Disable carrier**.

In your example the carrier price is correct with 10€ + 5€ = 15€, even though it displays CarrierA with 15€ (undesirable but doesn't create an issue with what the customer has to pay).

If you set it to **Disable carrier**, the checkout will not display any carrier and prevent you from completing the order.

**Comment by Hlavtox** [comment]
Yep, reopening, this is something new @paulnoelcholot, although related. :-) 👍

**Comment by paulnoelcholot** [comment]
Hello @Padamdam

My apologize, thank you @Padamdam and @Hlavtox :pray: 

I tested your issues in 8.1.4 with yours details I reproduce the problem as you can see :

https://www.loom.com/share/135f961bec974e5fa9010774e9596b30?sid=c535d052-03cb-4894-9140-55f94695292e

I'll add this to the backlog so it can be fixed.

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you

**Comment by boo-code** [comment]
Reproduced on 9.2.x with the configuration described, and the `getCarriersForOrder()` line in the report is one of four. `Carrier::checkDeliveryPriceByWeight()` is called with `$cart->getTotalWeight()` in every one of them:

| Where | What it decides |
| --- | --- |
| `Carrier::getCarriersForOrder()` | removes the carrier from the list offered for the cart |
| `Cart::isCarrierInRange()` | resets the requested carrier, so the package is priced with the default carrier instead |
| `Cart::getPackageShippingCostValue()`, carrier selection loop | skips the carrier while looking for the cheapest one in range |
| `Cart::getPackageShippingCostValue()`, final cost | adds `0` instead of the range price |

The second one explains the remark in the report that the delivery amount still looked right. It does not fail loudly: the requested carrier is silently dropped and the package is priced with `PS_CARRIER_DEFAULT`, which is why a single carrier's price shows up for the whole order rather than the sum.

The fourth is the clearest: the check sits one line above

```php
$shipping_cost += $carrier->getDeliveryPriceByWeight($this->getTotalWeight($product_list), $id_zone);
```

so the check weighs the cart while the price it guards weighs the package. `$order_total`, used by the price based branch of the same `if`, is already computed from `$product_list`.

Measured with two carriers 0-3 kg set to "Disable carrier" and two 2 kg products restricted one to each:

| | before | after |
| --- | --- | --- |
| `getCarriersForOrder()` | both carriers dropped | both kept |
| `getPackageList()` | one package, `carrier_list = [0]` | two packages, `[A]` and `[B]` |
| `getDeliveryOptionList()` | `[]` | one option per package |
| `getPackageShippingCost($carrier, true, $country, $package)` | `0.0` | `10.0` |

`getCarriersForOrder()` runs before packages exist, so there it weighs the products the carrier is allowed to ship, which bounds any package it can be handed. `Cart::isCarrierInRange()` gains an optional product list defaulting to the whole cart, so existing callers are unaffected.

Two related spots left alone on purpose: `Carrier::getAvailableCarrierList()` compares `$carrier->max_weight` against the whole cart weight, which is a different setting from the delivery ranges and deserves its own decision; and `PS_SHIPPING_FREE_WEIGHT` correctly stays cart-wide since it is an order level threshold.

Fix and an integration test building this exact scenario are prepared against 9.2.x. I am at the open-PR limit, so I will link the PR here once a slot frees.
